File: runners/loader.py
"""
Eval loader — reads a self-contained eval directory.

Each eval directory contains:
  PROMPT.md   — frontmatter metadata + ## System and ## Task sections
  graders.py  — define_graders() returning a list of grader dicts
  scaffold/   — (optional) starter files written to the agent workspace
"""

import importlib.util
import re
import sys
from dataclasses import dataclass, field
from pathlib import Path


@dataclass
class EvalDefinition:
    id: str
    name: str
    category: str
    path: Path
    system_prompt: str
    user_prompt: str
    graders: list[dict]
    scaffold: dict[str, str]               # relative_path -> content
    skills: list[str] = field(default_factory=list)   # skill names to fetch
    metadata: dict = field(default_factory=dict)


def load_eval(eval_config: dict, framework_root: Path) -> EvalDefinition:
    """
    Load an eval from its directory given a registry entry dict.

    eval_config: entry from config/evaluations.py::EVALUATIONS
    framework_root: absolute path to eval_framework/
    """
    eval_path = framework_root / eval_config["path"]
    if not eval_path.is_dir():
        raise FileNotFoundError(f"Eval directory not found: {eval_path}")

    system_prompt, user_prompt, meta = _parse_prompt_md(eval_path / "PROMPT.md")
    graders_module = _load_graders_module(eval_path / "graders.py")
    scaffold = _load_scaffold(eval_path / "scaffold")

    skills = [s.strip() for s in meta.get("skills", "").split(",") if s.strip()]

    return EvalDefinition(
        id=eval_config["id"],
        name=eval_config.get("name", meta.get("name", eval_config["id"])),
        category=eval_config.get("category", meta.get("category", "")),
        path=eval_path,
        system_prompt=system_prompt,
        user_prompt=user_prompt,
        graders=graders_module.define_graders(),
        scaffold=scaffold,
        skills=skills,
        metadata={
            "provider_name": meta.get("provider_name", "Auth0"),
            "provider_url":  meta.get("provider_url", "auth0.com"),
            "category":      eval_config.get("category", ""),
            "task_description": meta.get("task_description", eval_config.get("name", "")),
        },
    )


# ── PROMPT.md parser ──────────────────────────────────────────────────────────

def _parse_prompt_md(prompt_path: Path) -> tuple[str, str, dict]:
    """
    Parse PROMPT.md into (system_prompt, user_prompt, metadata_dict).

    Expected format:
      ---
      key: value
      ---

      ## System
      <system prompt text>

      ## Task
      <user/task prompt text>
    """
    if not prompt_path.exists():
        raise FileNotFoundError(f"PROMPT.md not found: {prompt_path}")

    text = prompt_path.read_text()

    # Extract YAML-ish frontmatter between --- delimiters
    meta: dict = {}
    front_match = re.match(r"^---\n(.*?)\n---\n", text, re.DOTALL)
    if front_match:
        for line in front_match.group(1).splitlines():
            if ":" in line:
                k, _, v = line.partition(":")
                meta[k.strip()] = v.strip()
        text = text[front_match.end():]

    # If ## System / ## Task sections exist, use them; otherwise treat the
    # whole body as the user prompt and use a minimal system prompt.
    system_match = re.search(r"^## System\s*\n(.*?)(?=^## |\Z)", text, re.DOTALL | re.MULTILINE)
    task_match   = re.search(r"^## Task\s*\n(.*?)(?=^## |\Z)",   text, re.DOTALL | re.MULTILINE)

    if system_match or task_match:
        system_prompt = system_match.group(1).strip() if system_match else ""
        user_prompt   = task_match.group(1).strip()   if task_match   else text.strip()
    else:
        system_prompt = "You are an expert iOS/Swift developer."
        user_prompt   = text.strip()

    return system_prompt, user_prompt, meta


# ── graders.py dynamic import ─────────────────────────────────────────────────

def _load_graders_module(graders_path: Path):
    """Dynamically import graders.py from an eval directory."""
    if not graders_path.exists():
        raise FileNotFoundError(f"graders.py not found: {graders_path}")

    spec = importlib.util.spec_from_file_location("_eval_graders", graders_path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    if not hasattr(module, "define_graders"):
        raise AttributeError(f"graders.py missing define_graders(): {graders_path}")

    return module


# ── scaffold loader ───────────────────────────────────────────────────────────

def _load_scaffold(scaffold_dir: Path) -> dict[str, str]:
    """
    Walk scaffold/ and return {relative_path: content} for every text file.
    Returns empty dict if scaffold/ does not exist.
    """
    if not scaffold_dir.is_dir():
        return {}

    files: dict[str, str] = {}
    for path in scaffold_dir.rglob("*"):
        if path.is_file():
            rel = str(path.relative_to(scaffold_dir))
            try:
                files[rel] = path.read_text()
            except Exception:
                pass
    return files

File: packages/eval/src/loader.ts
/**
 * Eval loader — reads a self-contained eval directory.
 *
 * Each eval directory contains:
 *   PROMPT.md   — frontmatter metadata + ## System and ## Task sections
 *   graders.ts  — defineGraders() returning a list of grader dicts
 *   scaffold/   — (optional) starter files written to the agent workspace
 */

import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs';
import { join, relative } from 'node:path';
import { pathToFileURL } from 'node:url';
import { EvalConfigError, EvalNotFoundError } from './errors.js';
import type { EvalDefinition, GraderDef } from './types/eval.js';

export { EvalDefinition, GraderDef } from './types/eval.js';

/** Minimal eval registration entry — the config ID, display name, category, and path. */
export interface EvalConfig {
  id: string;
  name: string;
  category: string;
  path: string;
}

/** Options for customising how the loader parses PROMPT.md. */
export interface LoadEvalOptions {
  /** Default baseline system prompt when no `## System` section is found. */
  defaultBaselineSystemPrompt?: string;
}

const FALLBACK_BASELINE_PROMPT =
  'Always prefer the official Auth0 SDK for the target framework. Do not use generic or third-party alternatives when an official Auth0 package exists.';

export async function loadEval(
  evalConfig: EvalConfig,
  frameworkRoot: string,
  options?: LoadEvalOptions,
): Promise<EvalDefinition> {
  const evalPath = join(frameworkRoot, evalConfig.path);
  if (!existsSync(evalPath) || !statSync(evalPath).isDirectory()) {
    throw new EvalNotFoundError(evalConfig.id);
  }

  const defaultBaselinePrompt = options?.defaultBaselineSystemPrompt ?? FALLBACK_BASELINE_PROMPT;
  const { baselineSystemPrompt, userPrompt, meta } = parsePromptMd(join(evalPath, 'PROMPT.md'), defaultBaselinePrompt);

  // system_default.md is always the agent system prompt — it contains the universal
  // identity, 8-step workflow, tool guidance, and behavior rules. Individual evals
  // do not override this; framework-specific hints belong in the ## Task section.
  const srcDefaultPath = join(frameworkRoot, 'src', 'prompts', 'system_default.md');
  const distDefaultPath = join(frameworkRoot, 'prompts', 'system_default.md');
  const defaultPath = existsSync(srcDefaultPath) ? srcDefaultPath : distDefaultPath;
  const agentSystemPrompt = existsSync(defaultPath) ? readFileSync(defaultPath, 'utf-8').trim() : '';

  const distRelPath = evalConfig.path.replace(/^src\//, '');
  const distGradersPath = join(frameworkRoot, 'dist', distRelPath, 'graders.js');
  const srcGradersPath = join(evalPath, 'graders.ts');
  const gradersPath = existsSync(distGradersPath) ? distGradersPath : srcGradersPath;
  const graders = await loadGraders(gradersPath);
  const scaffold = loadScaffold(join(evalPath, 'scaffold'));

  const skillsRaw = meta.skills ?? '';
  const skills = skillsRaw
    .split(',')
    .map((s: string) => s.trim())
    .filter(Boolean);

  const setupCommand = meta.setup_command || undefined;

  return {
    id: evalConfig.id,
    name: evalConfig.name ?? meta.name ?? evalConfig.id,
    category: evalConfig.category ?? meta.category ?? '',
    path: evalPath,
    baselineSystemPrompt,
    userPrompt,
    agentSystemPrompt,
    graders,
    scaffold,
    setupCommand,
    skills,
    metadata: {
      provider_name: meta.provider_name ?? 'Auth0',
      provider_url: meta.provider_url ?? 'auth0.com',
      category: evalConfig.category ?? '',
      task_description: meta.task_description ?? evalConfig.name ?? '',
    },
  };
}

// ── PROMPT.md parser ──────────────────────────────────────────────────────────

function parsePromptMd(
  promptPath: string,
  defaultBaselinePrompt: string,
): {
  baselineSystemPrompt: string;
  userPrompt: string;
  agentSystemPrompt: string;
  meta: Record<string, string>;
} {
  if (!existsSync(promptPath)) {
    throw new EvalConfigError('PROMPT.md not found', promptPath);
  }

  let text = readFileSync(promptPath, 'utf-8').replace(/\r\n/g, '\n');

  // Extract YAML-ish frontmatter between --- delimiters
  const meta: Record<string, string> = {};
  const frontMatch = text.match(/^---\n([\s\S]*?)\n---\n/);
  if (frontMatch && frontMatch[1]) {
    for (const line of frontMatch[1].split('\n')) {
      const colonIdx = line.indexOf(':');
      if (colonIdx !== -1) {
        const k = line.slice(0, colonIdx).trim();
        const v = line.slice(colonIdx + 1).trim();
        meta[k] = v;
      }
    }
    if (frontMatch[0]) {
      text = text.slice(frontMatch[0].length);
    }
  }

  const systemMatch = text.match(/^## System\s*\n([\s\S]*?)(?=^## |(?![\s\S]))/m);
  const agentSystemMatch = text.match(/^## Agent System\s*\n([\s\S]*?)(?=^## |(?![\s\S]))/m);
  const taskMatch = text.match(/^## Task\s*\n([\s\S]*?)(?=^## |(?![\s\S]))/m);

  const baselineSystemPrompt = systemMatch?.[1] ? systemMatch[1].trim() : defaultBaselinePrompt;
  const agentSystemPrompt = agentSystemMatch?.[1] ? agentSystemMatch[1].trim() : '';
  const userPrompt = taskMatch?.[1] ? taskMatch[1].trim() : text.trim();

  return { baselineSystemPrompt, userPrompt, agentSystemPrompt, meta };
}

// ── graders.ts dynamic import ─────────────────────────────────────────────────

async function loadGraders(gradersPath: string): Promise<GraderDef[]> {
  if (!existsSync(gradersPath)) {
    throw new EvalConfigError('graders file not found', gradersPath);
  }

  const mod = await import(pathToFileURL(gradersPath).href);
  if (typeof mod.defineGraders !== 'function') {
    throw new EvalConfigError('graders.ts missing defineGraders()', gradersPath);
  }

  return mod.defineGraders();
}

// ── scaffold loader ───────────────────────────────────────────────────────────

function loadScaffold(scaffoldDir: string): Record<string, string> {
  if (!existsSync(scaffoldDir) || !statSync(scaffoldDir).isDirectory()) {
    return {};
  }

  const files: Record<string, string> = {};
  walkDir(scaffoldDir, scaffoldDir, files);
  return files;
}

function walkDir(dir: string, root: string, files: Record<string, string>): void {
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const fullPath = join(dir, entry.name);
    if (entry.isDirectory()) {
      walkDir(fullPath, root, files);
    } else if (entry.isFile()) {
      const rel = relative(root, fullPath);
      try {
        files[rel] = readFileSync(fullPath, 'utf-8');
      } catch {
        // skip unreadable files
      }
    }
  }
}
